binaries/runtime/src/lib.rs

```diff
@@ -0,0 +1,187 @@
+    let mut operator_events_tx = HashMap::new();
+
+    for operator_config in &operators {
+        let (events_tx, events) = mpsc::channel(1);
+        let stop_publisher = publisher(
+            &node_id,
+            operator_config.id.clone(),
+            STOP_TOPIC.to_owned().into(),
+            communication.as_mut(),
+        )
+        .with_context(|| {
+            format!(
+                "failed to create stop publisher for operator {}",
+                operator_config.id
+            )
+        })?;
+        operator_stop_publishers.insert(operator_config.id.clone(), stop_publisher);
+
+        operator_events.insert(operator_config.id.clone(), ReceiverStream::new(events));
+        operator_events_tx.insert(operator_config.id.clone(), events_tx);
+    }
+
+    let operator_events = operator_events.map(|(id, event)| Event::Operator { id, event });
+    let node_id_clone = node_id.clone();
+    let stop_thread = std::thread::spawn(move || {
+        Builder::new_current_thread()
+            .enable_all()
+            .build()
+            .unwrap()
+            .block_on(run(
+                node_id_clone,
+                operator_events,
+                operator_stop_publishers,
+            ))
+            .unwrap();
+    });
+
+    for operator_config in &operators {
+        let events_tx = operator_events_tx.get(&operator_config.id).unwrap();
+        spawn_operator(
+            &node_id,
+            operator_config.clone(),
+            events_tx.clone(),
+            communication.as_mut(),
+        )
+        .wrap_err_with(|| format!("failed to init operator {}", operator_config.id))?;
+    }
+
+    stop_thread
+        .join()
+        .map_err(|err| eyre::eyre!("Stop thread failed with err: {err:#?}"))?;
+    Ok(())
+}
+
+async fn run(
+    node_id: NodeId,
+    mut events: impl Stream<Item = Event> + Unpin,
+    mut operator_stop_publishers: HashMap<OperatorId, Box<dyn Publisher>>,
+) -> eyre::Result<()> {
+    #[cfg(feature = "metrics")]
+    let _started = {
+        use dora_metrics::init_meter;
+        use opentelemetry::global;
+        use opentelemetry_system_metrics::init_process_observer;
+
+        let _started = init_meter();
+        let meter = global::meter(Box::leak(node_id.to_string().into_boxed_str()));
+        init_process_observer(meter);
+        _started
+    };
+
+    let mut stopped_operators = BTreeSet::new();
+
+    while let Some(event) = events.next().await {
+        match event {
+            Event::Operator { id, event } => {
+                match event {
+                    OperatorEvent::Error(err) => {
+                        bail!(err.wrap_err(format!("operator {id} failed")))
+                    }
+                    OperatorEvent::Panic(payload) => std::panic::resume_unwind(payload),
+                    OperatorEvent::Finished { reason } => {
+                        if let Some(stop_publisher) = operator_stop_publishers.remove(&id) {
```

Comment:
I think you accidentally removed support for the `ExplicitStopAll` handling introduced in #134 here. See https://github.com/dora-rs/dora/pull/134/files#diff-6a4df06248f886d7c9a17ad464fbd9aec8e6f6269b51946a5312f5e8c51274eeR113-R120.

Reply:
As I'm spawning the "stopping thread" in another thread than the main thread, I have to borrow the communication layer for static. And I'm not sure how to get around it.

Reply:
Oh actually we can, by replacing the `manual_stop_publisher` from a `FnOnce` closure to a `publisher`. This makes `communication` only borrowed for the time needed to create a `publisher`.